x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/SparseFileTracker.java

```diff
@@ -232,19 +231,28 @@ private List<Gap> doWaitForRange(ByteRange range, ByteRange subRange, ActionList
             assert targetRange.start == range.end() : targetRange;
             assert invariant();
 
-            assert ranges.containsAll(pendingRanges) : ranges + " vs " + pendingRanges;
             assert pendingRanges.stream().allMatch(Range::isPending) : pendingRanges;
             assert pendingRanges.size() != 1 || gaps.size() <= 1 : gaps;
-        }
 
-        // Pending ranges that needs to be filled before executing the listener
-        if (range.equals(subRange) == false) {
-            pendingRanges.removeIf(pendingRange -> (pendingRange.start < subRange.end() && subRange.start() < pendingRange.end) == false);
-            pendingRanges.sort(RANGE_START_COMPARATOR);
-        }
+            // Pending ranges that need to be filled before executing the listener
+            pendingRangesForListener = new ArrayList<>(pendingRanges);
+            if (range.equals(subRange) == false) {
+                pendingRangesForListener.removeIf(
+                    pendingRange -> (pendingRange.start < subRange.end() && subRange.start() < pendingRange.end) == false
+                );
+            }
 
-        subscribeToCompletionListeners(pendingRanges, subRange.end(), wrappedListener);
+            if (pendingRangesForListener.isEmpty()) {
+                // Do not return gaps to fill if the bytes to read are available and the listener can be invoked immediately.
+                gaps.clear();
```

Comment:
I worry somewhat about the approach here in that we'd get here over and over and never fill these gaps, thus doing lots of work for every read.

I wonder if there are alternatives? Perhaps we can let the fast path read check the ranges too. Or do a single read from object store to fill all the gaps in one go. Doing the full read to ensure that `complete` is updated could make sense. We can also read the gaps prior to the read-range to accomplish this.

Reply:
Hi @henningandersen , interesting thought-provoking comment!

> we'd get here over and over and never fill these gaps

That is possible, but seems unlikely to me that so many reads for so many disjoint filled ranges will happen, without a single unavailable byte (that would ultimately then fill gaps), that it would be of considerable performance impact. I am wondering if we are focusing on an edge case? Or maybe I am mistaken in thinking it would be rare?

> Perhaps we can let the fast path read check the ranges too.

Yes that's possible, but we'll still need to do it in a synchronized `ranges` block. We will, however, not do the unnecessary work of keeping the gaps to fill; we'll just check if all read bytes are available. But I am concerned we'll now impact the fast path much more performance-wise, assuming it's the most frequent code path, as it will do much more than simply checking the `complete` variable.

> Or do a single read from object store to fill all the gaps in one go.

I think the combination of @ywangd 's https://github.com/elastic/elasticsearch/pull/110855 PR and my in-review https://github.com/elastic/elasticsearch-serverless/pull/2382 PR is already achieving that. Meaning, if we're fine with paying the price to go the object store to fill the gaps, it will usually happen in one go. I was not sure we're fine with this, because of the existence of the ticket (?), but if we're fine with going to the blob store, we could think about dropping this PR/ticket.

Reply:
> But I am concerned we'll now impact the fast path much more performance-wise, assuming it's the most frequent code path, as it will do much more than simply checking the complete variable.

I think changes proposed here could already impact performance of the fast-path? If a file (or same portion of the file) is read multiple times and the file is in the middle of a region, the `complete` field will be `0` which forces the code to go through the slower path, i.e. computing the gaps under synchronization but then dropping them because the ranges are already available. Maybe this is what you mean by it being an edge case. But if this is indeed an edge case, doesn't that also mean the changes here save round-trips to object store for edge cases only? Additionally, since the ongoing work will ensure at most one request to the object store to fill the entire region, I am not entirely sure whether this is worth optimizing.

Reply:
Hi @ywangd , thanks for the feedback as well!

> computing the gaps under synchronization but then dropping them because the ranges are already available.

Indeed, but I think this is far better/faster than doing a roundtrip to the blob store (which requires much more time). I can see the reasoning that if we have like a gazillion such queries, that do not fill gaps, then the synchronized block's processing time overall might offset/overshadow the potential roundtrip time to the blob store to fill the gaps, but that's exactly the "edge case" I was thinking: by "edge case" I meant the following conditions need to apply for the synchronized block to become more expensive than the blob store roundtrip: (a) multiple queries that always try to read from already filled regions, (b) the rangeToWrite should have many disjoint holes. Plus, I think the normal case will be that some query will request at least one unavailable byte and fill the region's gaps.

At least that's what I thought was the reasoning behind the mere existence of the ticket. To save us an expensive round-trip to the blob store for a couple of queries that can read from already filled ranges. And only do the round-trip when a query needs at least one unavailable byte.

However, if you both agree that it's better to do the roundtrip to the blob store to fill the gaps either way, I think that's also fine, since this will be a single blob store roundtrip with the recent PRs, it will not be that expensive. So we can close the ticket and this PR. I'll do that unless there's more feedback or alternatives.

BTW, this reminds me of my argument for having a design phase for tickets: such discussions earlier would have saved some of the implementation time :)

Reply:
I agree with the concerns about performance as well as the utility to implement this now.

I initially opened the JIRA ticket as I've been puzzled by the fact we were reaching out to the object store to fill a complete range whereas we were reading the exact same bytes over and over in a file. It is possible that the bytes in question were just header/codec/footer, in which case we should maybe have different solution to optimize the reads. Otherwise, reaching out to the pending ranges as proposed in this PR will have a quite important performance impact (I remember just instanciating the List in the synchronize block had an impact) so we maybe want to look at alternatives or maybe just drop this ticket for now (since we now fetch multiple gaps at once, which should also help)

Reply:
> (I remember just instanciating the List in the synchronize block had an impact)

That indeed sounds concerning!

Since you also agree, I think it may make more sense to drop this ticket for now until we see it's truly a problem we'd like to optimize for and find a good alternative to do it.

Reply:
> I think the combination of @ywangd 's https://github.com/elastic/elasticsearch/pull/110855 PR and my in-review https://github.com/elastic/elasticsearch-serverless/pull/2382 PR is already achieving that

Thanks for the links, looks like it is already done (the other PR is not really necessary for this purpose).

I think we could eventually want a sort of cleanup process that ensures that half-filled regions are filled. But the current functionality (without this PR, but including #110855) would automatically fill used regions and we'd expect unused ones to age out, so perhaps that is not even necessary if we avoid the work in this PR.

Thus not merging this PR is a feature :slightly_smiling_face: (until we realize otherwise ofc).

I agree to closing this PR.

Reply:
> I think it may make more sense to drop this ticket for now until we see it's truly a problem we'd like to optimize for and find a good alternative to do it.

:+1: too

Reply:
> BTW, this reminds me of my argument for having a design phase for tickets: such discussions earlier would have saved some of the implementation time :)

More than a design phase, I think we haven't discussed enough in team about some of the tickets that were brought into the iteration.

Reply:
> (the other PR is not really necessary for this purpose).

It just helps on tops of Yang's PR to have fewer round trips to the blob store.

Sounds good then, I'll be closing this, thanks for the discussion!